internal/provider/resource_gitlab_project_protected_environment.go

```diff
@@ -0,0 +1,205 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+// https://docs.gitlab.com/ee/ci/environments/protected_environments.html
+var _ = registerResource("gitlab_project_protected_environment", func() *schema.Resource {
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_project_protected_environment`" + ` resource you to create and manage a protected environment in your GitLab project`,
+
+		CreateContext: resourceGitlabProjectProtectedEnvironmentCreate,
+		ReadContext:   resourceGitlabProjectProtectedEnvironmentRead,
+		DeleteContext: resourceGitlabProjectProtectedEnvironmentDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Description:  "The ID or full path of the project which the protected environment is created against.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"environment": {
+				Description:  "The name of the environment.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+		// Placeholder attribute till GitLab makes approval count live https://gitlab.com/gitlab-org/gitlab/-/issues/343864
+		// 	"required_approval_count": {
+		// 		Description:  "The number of approvals required to deploy to this environment. This is part of Deployment Approvals, which isn't yet available for use.",
+		// 		Type:         schema.TypeString,
+		// 		ForceNew:     true,
+		// 		Required:     false,
+		// 		ValidateFunc: validation.StringIsNotEmpty,
+		// },
+			"deploy_access_levels": {
+				Description: "Array of access levels allowed to deploy, with each described by a hash.",
+				Type:        schema.TypeList,
+				ForceNew:    true,
+				Required:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"access_level": {
+							Description:  fmt.Sprintf("Levels of access required to deploy to this protected environment: %s", renderValueListForDocs(validProtectedEnvironmentDeploymentLevelNames)),
```

Comment:
```suggestion
							Description:  fmt.Sprintf("Levels of access required to deploy to this protected environment. Valid values are %s.", renderValueListForDocs(validProtectedEnvironmentDeploymentLevelNames)),
```

This `Valid values are %s.` pattern is used in almost every resource ;)